packages/lesswrong/server/scripts/migrateCollections.ts

```diff
@@ -0,0 +1,181 @@
+      collection: collection.getMongoCollection() as unknown as CollectionBase<DbObject>,
+      batchSize,
+      filter: makeBatchFilter(createdSince),
+      callback: async (documents: DbObject[]) => {
+        console.log(`.........Migrating ${collection.getName()} documents ${count}-${count + documents.length}`);
+        count += batchSize;
+        const query = new InsertQuery(table, documents.map(formatter), {}, {conflictStrategy: "ignore"});
+        const compiled = query.compile();
+        try {
+          await sql.none(compiled.sql, compiled.args);
+        } catch (e) {
+          throw new Error(`Error importing document batch for collection ${collection.getName()}: ${e.message}`);
+        }
+      },
+    });
+  }
+}
+
+export const migrateCollections = async (collectionNames: CollectionNameString[]) => {
+  console.log(`=== Migrating collections '${collectionNames}' from Mongo to Postgres ===`);
+
+  console.log("...Preparing collections");
+  const collections = collectionNames.map(getCollection) as unknown as SwitchingCollection<DbObject>[];
+  for (const [index, collection] of collections.entries()) {
+    if (!collection) {
+      throw new Error(`Invalid collection name '${collectionNames[index]}'`);
+    }
+
+    if (!(collection instanceof SwitchingCollection)) {
+      throw new Error(`Collection '${collectionNames[index]}' is not a SwitchingCollection`);
+    }
+
+    collection.setTargets("mongo", "mongo");
+  }
+
+  console.log("...Beginning SQL transaction");
+  const sql = getSqlClientOrThrow();
+  try {
+    await sql.tx(async (transaction) => {
+      // Create the tables
+      await createTables(transaction, collections);
+      await createIndexes(transaction, collections);
+      await setLogged(transaction, collections, false);
+
+      // Copy the initial data - this can take a *long* time depending on the collection
+      const copyStart = new Date();
+      await copyData(transaction, collections, 1);
+      await setLogged(transaction, collections, true);
+
+      // Copy anything that was inserted during the first copy - this should be ~instant
+      const copyStart2 = new Date();
+      await copyData(transaction, collections, 2, copyStart);
+
+      // Start writing to Postgres
+      for (const collection of collections) {
+        collection.setTargets("mongo", "pg");
+      }
+
+      // Once more for luck...
+      await copyData(transaction, collections, 3, copyStart2);
+
+      // Fully move to Postgres and write the lock in case the server restarts at some point
+      await Promise.all(collections.map((collection) => {
+        collection.setTargets("pg", "pg");
+        return collection.writeToLock();
+      }));
+
+      console.log("\ud83c\udfc1 Done!");
+    });
+  } catch (e) {
+    for (const collection of collections) {
+      collection.setTargets("mongo", "mongo");
+    }
+    console.error("Error:", e.message);
+    console.log("\ud83d\uded1 Migration aborted!");
+  }
```

Comment:
I think it would be good to have the ability to manually switch back to mongo quickly after the fact, e.g. if we see that 5xx errors go way up we should be able to revert it in under 30 seconds to minimise the amount of data loss. Maybe this should even be automatic although that would just introduce a new risk

Reply:
That's a good point - this should be solvable by just continuing to poll even after we switch to postgres.